File: ios/Sources/RuvoiOS/Features/Weather/WeatherService.swift
import Foundation
import CoreLocation

// MARK: – Models

struct WeatherCondition {
    let emoji: String
    let condition: String
    let temperature: String
    let feelsLike: String
    let humidity: Int
    let windSpeed: Double
    let isGoodForRun: Bool
    let recommendation: String
    let tipTitle: String
    let tip: String
}

// MARK: – OpenWeather API Response

private struct OWResponse: Decodable {
    struct WeatherItem: Decodable { let id: Int; let description: String }
    struct Main: Decodable {
        let temp: Double; let feels_like: Double; let humidity: Int
    }
    struct Wind: Decodable { let speed: Double }
    let weather: [WeatherItem]
    let main: Main
    let wind: Wind
}

// MARK: – Service

@MainActor
final class WeatherService: ObservableObject {
    @Published var condition: WeatherCondition? = nil
    @Published var isLoading = false

    private let locationManager = CLLocationManager()
    private var locationDelegate: OneShotLocationDelegate?

    func fetchIfNeeded() async {
        guard condition == nil else { return }
        await fetch()
    }

    func fetch() async {
        isLoading = true
        defer { isLoading = false }
        guard let loc = await requestLocation() else { return }
        let apiKey = Bundle.main.object(forInfoDictionaryKey: "OPENWEATHER_API_KEY") as? String ?? ""
        guard !apiKey.isEmpty else { return }

        var comps = URLComponents(string: "https://api.openweathermap.org/data/2.5/weather")!
        comps.queryItems = [
            .init(name: "lat",   value: "\(loc.coordinate.latitude)"),
            .init(name: "lon",   value: "\(loc.coordinate.longitude)"),
            .init(name: "units", value: "metric"),
            .init(name: "appid", value: apiKey),
        ]
        guard let url = comps.url,
              let (data, _) = try? await URLSession.shared.data(from: url),
              let response = try? JSONDecoder().decode(OWResponse.self, from: data) else { return }

        condition = buildCondition(from: response)
    }

    private func requestLocation() async -> CLLocation? {
        locationManager.requestWhenInUseAuthorization()
        return await withCheckedContinuation { cont in
            let delegate = OneShotLocationDelegate { [weak self] loc in
                self?.locationDelegate = nil
                cont.resume(returning: loc)
            }
            locationDelegate = delegate
            locationManager.delegate = delegate
            locationManager.requestLocation()
        }
    }

    private func buildCondition(from r: OWResponse) -> WeatherCondition {
        let id   = r.weather.first?.id ?? 800
        let desc = (r.weather.first?.description ?? "Clear").capitalized
        let temp = r.main.temp; let feels = r.main.feels_like
        let humidity = r.main.humidity; let wind = r.wind.speed

        let isRain   = (200...599).contains(id)
        let isStorm  = (200...299).contains(id)
        let isSnow   = (600...699).contains(id)
        let isClear  = id == 800

        let emoji: String
        switch id {
        case 200...299: emoji = "⛈️"
        case 300...399: emoji = "🌦️"
        case 500...599: emoji = "🌧️"
        case 600...699: emoji = "❄️"
        case 800:       emoji = "☀️"
        case 801...802: emoji = "⛅"
        default:        emoji = "🌤️"
        }

        let isGood = !isStorm && !isSnow && (-5.0...35.0).contains(temp) && humidity < 90 && wind < 15

        let rec: String
        switch true {
        case isStorm:    rec = "Stay indoors — lightning risk is high."
        case isSnow:     rec = "Slippery surfaces. Consider a treadmill today."
        case isRain && temp > 15: rec = "Light rain is fine! Wear a cap and stay visible."
        case isRain:     rec = "Cold rain — waterproof jacket and warm layers."
        case temp > 32:  rec = "Extreme heat — run early morning, stay hydrated."
        case temp > 28:  rec = "Hot: slow down 10–15%, hydrate every 20 min."
        case temp < 5:   rec = "Cold run: layer up, warm up indoors first."
        case wind > 10:  rec = "Strong wind — run into it first for easier return."
        case isClear && (10.0...22.0).contains(temp): rec = "Perfect conditions today! Great day for a PR. 🏆"
        default:         rec = "Good conditions. Enjoy your run!"
        }

        let (tipTitle, tip): (String, String)
        switch true {
        case temp > 28:  (tipTitle, tip) = ("Heat Protocol", "Pre-cool with cold water, wear light colors, cut pace 10–15%.")
        case temp < 5:   (tipTitle, tip) = ("Cold Running", "Cover extremities. Expect 20–30s/km slower pace.")
        case isRain:     (tipTitle, tip) = ("Wet Weather", "Shorten stride, avoid puddles, anti-chafe on feet.")
        case wind > 8:   (tipTitle, tip) = ("Wind Tips", "Start into the headwind, benefit on return.")
        case humidity > 75: (tipTitle, tip) = ("Humidity", "High humidity feels warmer. Hydrate more.")
        default:         (tipTitle, tip) = ("Optimal Day", "Ideal temperature range. Good day to push your limits.")
        }

        return WeatherCondition(
            emoji: emoji, condition: desc,
            temperature: String(format: "%.0f°C", temp),
            feelsLike: String(format: "Feels %.0f°C", feels),
            humidity: humidity, windSpeed: wind,
            isGoodForRun: isGood,
            recommendation: rec, tipTitle: tipTitle, tip: tip
        )
    }
}

// One-shot CLLocationManagerDelegate (nonisolated since CLLocationManager callbacks run on main thread)
final class OneShotLocationDelegate: NSObject, CLLocationManagerDelegate {
    private let completion: (CLLocation?) -> Void
    private var done = false
    init(completion: @escaping (CLLocation?) -> Void) { self.completion = completion }
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard !done else { return }; done = true
        completion(locations.first)
    }
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        guard !done else { return }; done = true
        completion(nil)
    }
}

File: android/app/src/main/java/com/ruvo/app/features/weather/WeatherService.kt
package com.ruvo.app.features.weather

import android.annotation.SuppressLint
import android.content.Context
import android.location.Location
import com.google.android.gms.location.LocationServices
import com.ruvo.app.BuildConfig
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import javax.inject.Inject
import javax.inject.Singleton

@Serializable
data class WeatherResponse(
    val weather: List<WeatherItem> = emptyList(),
    val main: MainData = MainData(),
    val wind: WindData = WindData(),
    val rain: RainData? = null,
    val name: String = "",
)

@Serializable
data class WeatherItem(val id: Int = 0, val main: String = "", val description: String = "", val icon: String = "")

@Serializable
data class MainData(
    val temp: Double = 0.0,
    @SerialName("feels_like") val feelsLike: Double = 0.0,
    val humidity: Int = 0,
)

@Serializable
data class WindData(val speed: Double = 0.0, val deg: Int = 0)

@Serializable
data class RainData(@SerialName("1h") val oneHour: Double? = null)

data class RunWeatherAdvice(
    val emoji: String,
    val condition: String,
    val temperature: String,
    val recommendation: String,
    val isGoodForRun: Boolean,
    val tipTitle: String,
    val tip: String,
)

@Singleton
class WeatherService @Inject constructor(
    @ApplicationContext private val context: Context,
) {
    private val client = OkHttpClient()
    private val json   = Json { ignoreUnknownKeys = true }

    @SuppressLint("MissingPermission")
    suspend fun fetchCurrentWeather(): WeatherResponse? = withContext(Dispatchers.IO) {
        try {
            val fused = LocationServices.getFusedLocationProviderClient(context)
            val loc: Location = fused.lastLocation.await() ?: return@withContext null
            val apiKey = BuildConfig.OPENWEATHER_API_KEY
            if (apiKey.isBlank()) return@withContext null
            val url = "https://api.openweathermap.org/data/2.5/weather?lat=${loc.latitude}&lon=${loc.longitude}&units=metric&appid=$apiKey"
            val req = Request.Builder().url(url).build()
            val body = client.newCall(req).execute().body?.string() ?: return@withContext null
            json.decodeFromString<WeatherResponse>(body)
        } catch (_: Exception) { null }
    }

    fun buildAdvice(weather: WeatherResponse): RunWeatherAdvice {
        val temp = weather.main.temp
        val humidity = weather.main.humidity
        val windSpeed = weather.wind.speed
        val conditionId = weather.weather.firstOrNull()?.id ?: 800
        val description = weather.weather.firstOrNull()?.description?.replaceFirstChar { it.uppercase() } ?: "Clear"

        val isRaining = conditionId in 200..599
        val isStormy  = conditionId in 200..299 || conditionId in 900..902
        val isSnow    = conditionId in 600..699
        val isClear   = conditionId in 800..802

        val emoji = when {
            isStormy  -> "⛈️"
            isRaining -> "🌧️"
            isSnow    -> "❄️"
            conditionId == 800 -> "☀️"
            conditionId in 801..802 -> "⛅"
            else -> "🌤️"
        }

        val isGoodForRun = !isStormy && !isSnow && temp in -5.0..35.0 && humidity < 90 && windSpeed < 15

        val recommendation = when {
            isStormy  -> "Stay indoors — lightning risk is high."
            isSnow    -> "Run with caution on slippery surfaces. Consider treadmill."
            isRaining -> if (temp > 15) "Light rain is fine! Wear a cap and stay visible." else "Cold rain: waterproof jacket and warm layers."
            temp > 32 -> "Extreme heat — run early morning, stay hydrated."
            temp > 28 -> "Hot weather — slow down 10–15%, hydrate every 20 min."
            temp < 5  -> "Cold run: layer up, warm up indoors first."
            windSpeed > 10 -> "Strong wind — run into the wind first for easier return."
            isClear && temp in 10.0..22.0 -> "Perfect running conditions today! Great day for a PR."
            else -> "Good conditions. Enjoy your run!"
        }

        val (tipTitle, tip) = when {
            temp > 28  -> "Heat Protocol" to "Pre-cool with cold water, wear light colors, cut pace by 10–15%."
            temp < 5   -> "Cold Running"  to "Cover extremities. Expect 20–30s/km slower pace. Breathe through nose."
            isRaining  -> "Wet Weather"   to "Shorten stride, avoid puddles, anti-chafe cream on feet."
            windSpeed > 8 -> "Wind Tips"  to "Start into the headwind, benefit from tailwind on return."
            humidity > 75 -> "Humidity"   to "High humidity feels warmer. Hydrate more, expect slower pace."
            else        -> "Optimal Day"  to "Ideal temperature range. Good day to push your limits."
        }

        return RunWeatherAdvice(
            emoji = emoji,
            condition = description,
            temperature = String.format("%.0f°C (feels %.0f°C)", temp, weather.main.feelsLike),
            recommendation = recommendation,
            isGoodForRun = isGoodForRun,
            tipTitle = tipTitle,
            tip = tip,
        )
    }
}
